agent/src/cli/evaluate-autocomplete/Queries.ts

```diff
@@ -0,0 +1,79 @@
+import * as fspromises from 'fs/promises'
+import * as path from 'path'
+
+import Parser, { Query } from 'web-tree-sitter'
+
+import { SupportedLanguage } from '../../../../vscode/src/completions/tree-sitter/grammars'
+
+export type QueryName = 'context'
+
+/**
+ * Queries manages compilation of tree-sitter queries from a directory layout.
+ *
+ * Queries are written in a file structure like this: `LANGUAGE/QUERY_NAME.scm`
+ *
+ * This class caches compilation of queries so that we only read each query once from disk.
+ */
+export class Queries {
+    private cache: CompiledQuery[] = []
+    constructor(private queriesDirectory: string) {}
+    public async loadQuery(parser: Parser, language: SupportedLanguage, name: QueryName): Promise<Query | undefined> {
+        const fromCache = this.cache.find(compiled => compiled.language === language && compiled.queryName === name)
+        if (fromCache) {
+            return fromCache.compiledQuery
+        }
+        return this.compileQuery(parser, language, name)
```

Comment:
`?.` and short-circuit evaluation is more succinct and idiomatic.

```suggestion
        const cachedQuery = this.cache.find(compiled => compiled.language === language && compiled.queryName === name)?.compiledQuery
        return cachedQuery || this.compileQuery(parser, language, name)
```

Reply:
Either way works for me. I like the current solution because I find it easy to read and easier to extend. Operators like `?` tend to become unwieldy when you have more complex expressions or want to do things with intermediary values. The current structure makes it easy to add new statements or extract other properties.